File: domain/case.ts
/**
 * domain/case.ts
 *
 * Case is the pure domain model for the top-level administrative and
 * accountability container for platform work. A Case has a single
 * accountable human owner, per the Human Accountability principle: the
 * platform may assist, recommend, and organize, but it never removes a
 * human from final ownership of a case.
 *
 * Mirrors contracts/domain/case.schema.json field-for-field.
 *
 * Domain Purity: standard runtime and sibling domain modules only.
 */

import {
  Status,
  freezeRefs,
  freezeTags,
  newId,
  now,
  type Id,
  type Metadata,
  type Provenance,
  type Reference,
  type Timestamp,
  EMPTY_METADATA,
} from "./common.js";
import { ClassificationLevel } from "./classification_level.js";
import { DomainInvariantError } from "./common.js";

export enum CasePriority {
  Low = "low",
  Normal = "normal",
  High = "high",
  Critical = "critical",
}

export interface CaseInit {
  readonly title: string;
  readonly owner: string;
  readonly description?: string;
  readonly priority?: CasePriority;
  readonly openedAt?: Timestamp;
  readonly closedAt?: Timestamp;
  readonly status?: Status;
  readonly classification?: ClassificationLevel;
  readonly identityRefs?: ReadonlyArray<Reference>;
  readonly investigationRefs?: ReadonlyArray<Reference>;
  readonly provenance: Provenance;
  readonly tags?: ReadonlyArray<string>;
  readonly metadata?: Metadata;
}

/** Case is an immutable accountability container that groups one or more Investigations. */
export class Case {
  readonly id: Id;
  readonly version: string;
  readonly createdAt: Timestamp;
  readonly updatedAt: Timestamp;
  readonly status: Status;
  readonly classification: ClassificationLevel;
  readonly title: string;
  readonly description?: string;
  readonly priority: CasePriority;
  readonly owner: string;
  readonly openedAt: Timestamp;
  readonly closedAt?: Timestamp;
  readonly identityRefs: ReadonlyArray<Reference>;
  readonly investigationRefs: ReadonlyArray<Reference>;
  readonly provenance: Provenance;
  readonly tags: ReadonlyArray<string>;
  readonly metadata: Metadata;

  private constructor(id: Id, version: string, createdAt: Timestamp, updatedAt: Timestamp, init: CaseInit) {
    if (init.title.trim().length === 0) {
      throw new DomainInvariantError("Case.title must not be empty.");
    }
    if (init.owner.trim().length === 0) {
      throw new DomainInvariantError("Case.owner must not be empty: every case must have an accountable human owner.");
    }
    this.id = id;
    this.version = version;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
    this.status = init.status ?? Status.Active;
    this.classification = init.classification ?? ClassificationLevel.Confidential;
    this.title = init.title;
    this.description = init.description;
    this.priority = init.priority ?? CasePriority.Normal;
    this.owner = init.owner;
    this.openedAt = init.openedAt ?? now();
    this.closedAt = init.closedAt;
    this.identityRefs = freezeRefs(init.identityRefs);
    this.investigationRefs = freezeRefs(init.investigationRefs);
    this.provenance = init.provenance;
    this.tags = freezeTags(init.tags);
    this.metadata = init.metadata ?? EMPTY_METADATA;
    Object.freeze(this);
  }

  static open(init: CaseInit): Case {
    const timestamp = now();
    return new Case(newId(), "1.0.0", timestamp, timestamp, init);
  }

  /** Returns a new, closed Case. Throws if the case is already closed. */
  close(closedAt: Timestamp = now()): Case {
    if (this.status === Status.Archived) {
      throw new DomainInvariantError("Case is already closed/archived.");
    }
    return new Case(this.id, this.version, this.createdAt, now(), {
      title: this.title,
      owner: this.owner,
      description: this.description,
      priority: this.priority,
      openedAt: this.openedAt,
      closedAt,
      status: Status.Archived,
      classification: this.classification,
      identityRefs: this.identityRefs,
      investigationRefs: this.investigationRefs,
      provenance: this.provenance,
      tags: this.tags,
      metadata: this.metadata,
    });
  }

  /** Returns a new Case with the given investigation reference attached. */
  withInvestigationRef(investigationRef: Reference): Case {
    if (this.investigationRefs.some((ref) => ref.refId === investigationRef.refId)) {
      return this;
    }
    return new Case(this.id, this.version, this.createdAt, now(), {
      title: this.title,
      owner: this.owner,
      description: this.description,
      priority: this.priority,
      openedAt: this.openedAt,
      closedAt: this.closedAt,
      status: this.status,
      classification: this.classification,
      identityRefs: this.identityRefs,
      investigationRefs: [...this.investigationRefs, investigationRef],
      provenance: this.provenance,
      tags: this.tags,
      metadata: this.metadata,
    });
  }
}

File: domain/case.py
"""
domain/case.py

Case domain model - container for related investigations.
Mirrors contracts/domain/case.schema.json field-for-field.
"""
from dataclasses import dataclass, field
from enum import Enum

from .classification_level import ClassificationLevel, ClassificationLevelValue
from .common import (
    EMPTY_METADATA,
    Metadata,
    Provenance,
    Reference,
    Status,
    freeze_refs,
    new_id,
    utc_now,
)


class CasePriority(str, Enum):
    """Priority levels for cases."""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


@dataclass(frozen=True)
class Case:
    """
    Case is a container for related investigations.
    
    Per the Human Accountability principle, every Case has a named
    human owner and requires human review for closure.
    """
    id: str
    version: str
    created_at: str
    updated_at: str
    title: str
    description: str
    status: Status
    classification: ClassificationLevel
    priority: CasePriority
    owner: str
    investigation_refs: frozenset = field(default_factory=frozenset)
    report_refs: frozenset = field(default_factory=frozenset)
    tags: frozenset = field(default_factory=frozenset)
    metadata: Metadata = field(default_factory=lambda: dict(EMPTY_METADATA))

    @classmethod
    def create(
        cls,
        title: str,
        description: str,
        owner: str,
        provenance: Provenance,
        investigation_refs: list[Reference] | None = None,
        report_refs: list[Reference] | None = None,
        status: Status = Status.DRAFT,
        classification: ClassificationLevelValue = ClassificationLevelValue.INTERNAL,
        priority: CasePriority = CasePriority.MEDIUM,
        tags: list[str] | None = None,
        metadata: Metadata | None = None,
    ) -> "Case":
        """Create a new Case."""
        if not owner or not owner.strip():
            raise ValueError("Case.owner must not be empty: every case requires a human owner.")
        
        timestamp = utc_now()
        return cls(
            id=new_id(),
            version="1.0.0",
            created_at=timestamp,
            updated_at=timestamp,
            title=title,
            description=description,
            status=status,
            classification=ClassificationLevel(classification),
            priority=priority,
            owner=owner,
            investigation_refs=frozenset(investigation_refs) if investigation_refs else frozenset(),
            report_refs=frozenset(report_refs) if report_refs else frozenset(),
            tags=frozenset(tags) if tags else frozenset(),
            metadata=dict(metadata) if metadata else dict(EMPTY_METADATA),
        )

    def with_investigation_ref(self, ref: Reference) -> "Case":
        """Return a new Case with the given investigation reference added."""
        if ref in self.investigation_refs:
            return self
        return Case(
            id=self.id,
            version=self.version,
            created_at=self.created_at,
            updated_at=utc_now(),
            title=self.title,
            description=self.description,
            status=self.status,
            classification=self.classification,
            priority=self.priority,
            owner=self.owner,
            investigation_refs=self.investigation_refs | {ref},
            report_refs=self.report_refs,
            tags=self.tags,
            metadata=self.metadata,
        )

    def with_report_ref(self, ref: Reference) -> "Case":
        """Return a new Case with the given report reference added."""
        if ref in self.report_refs:
            return self
        return Case(
            id=self.id,
            version=self.version,
            created_at=self.created_at,
            updated_at=utc_now(),
            title=self.title,
            description=self.description,
            status=self.status,
            classification=self.classification,
            priority=self.priority,
            owner=self.owner,
            investigation_refs=self.investigation_refs,
            report_refs=self.report_refs | {ref},
            tags=self.tags,
            metadata=self.metadata,
        )


__all__ = [
    "Case",
    "CasePriority",
]
